From bc7378fe69ed1a2a7c7719193a6748e6adae5cf0 Mon Sep 17 00:00:00 2001
From: Aniket Randive <quic_arandive@quicinc.com>
Date: Mon, 7 Apr 2025 18:55:54 +0530
Subject: [PATCH 07/12] PENDING: soc: qcom: ccu-qup: Add read and poll API
 support

In asynchronous mode for write transfers, upon completion
of the actual transfer, the CCU QUP driver will notify the
client using a poll notification. The client can then
invoke the read API to retrieve CCU QUP client data, which
includes the transfer result, frame cookie, GPII number,
SE number, and other relevant information.

Change-Id: Ib4a9c4e576c88e7c459d57895466083e09bcf547
Signed-off-by: Aniket Randive <quic_arandive@quicinc.com>
---
 drivers/soc/qcom/qcom_ccu_qup.c       | 123 ++++++++++++++++++++++++++
 include/linux/soc/qcom/qcom_ccu_qup.h |  17 +++-
 2 files changed, 138 insertions(+), 2 deletions(-)

diff --git a/drivers/soc/qcom/qcom_ccu_qup.c b/drivers/soc/qcom/qcom_ccu_qup.c
index 472e86f162b1..f02773d047f0 100644
--- a/drivers/soc/qcom/qcom_ccu_qup.c
+++ b/drivers/soc/qcom/qcom_ccu_qup.c
@@ -14,14 +14,17 @@
 #include <linux/irq.h>
 #include <linux/kdev_t.h>
 #include <linux/kernel.h>
+#include <linux/list.h>
 #include <linux/module.h>
 #include <linux/of.h>
 #include <linux/of_platform.h>
 #include <linux/platform_device.h>
+#include <linux/poll.h>
 #include <linux/soc/qcom/qcom_ccu_qup.h>
 #include <linux/spinlock.h>
 #include <linux/types.h>
 #include <linux/uaccess.h>
+#include <linux/wait.h>
 
 #define GPII_0_IRQ		0
 #define GPII_14_IRQ		19
@@ -29,13 +32,128 @@
 #define CCU_SYNC_TIMERS		20
 #define CCU_MAX_GPII_PER_SE	3
 
+struct ccu_qup *ccud;
+
 static const struct file_operations ccu_fops = {
 	.owner = THIS_MODULE,
 	.write = ccu_config_write,
+	.read = ccu_response_read,
 	.open = ccu_open,
+	.poll = ccu_poll,
 	.release = ccu_release,
 };
 
+/*
+ * ccu_response_read - User space client use read api to read
+ *                     the notification information once transfer
+ *                     is completed in Asynchrnous mode for write
+ *                     transfer.
+ * @filp: file pointer of CCU QUP device.
+ * @buf: Data buffer pointer passed to user space which is of type
+ *       “struct ccu_qup_client_data”.
+ * @len:  Represents transfer length of the transaction
+ * @pos: file pointer position
+ *
+ * Returns a number of bytes actually read in case of success and
+ * failure code in case of any failures.
+ */
+static ssize_t ccu_response_read(struct file *filp, char __user *buf,
+				 size_t count, loff_t *f_pos)
+{
+	struct ccu_qup *ccu;
+	struct ccu_response_data *entry;
+	int ret;
+
+	if (!filp || !filp->private_data)
+		return -EINVAL;
+
+	ccu = filp->private_data;
+	if (list_empty(&ccu->response_list)) {
+		dev_err(ccu->dev, "No rx response data avilable\n");
+		return -EAGAIN;
+	}
+
+	entry = list_last_entry(&ccu->response_list, struct ccu_response_data, list);
+	if (entry) {
+		ret = copy_to_user(buf, entry->res_buf,
+				   sizeof(struct ccu_xfer_result));
+		if (ret) {
+			dev_err(ccu->dev, "copy to user failed\n");
+			return -EAGAIN;
+		}
+		list_del(&entry->list);
+		kfree(entry->res_buf);
+		kfree(entry);
+		atomic_dec(&ccu->rx_avail);
+
+		return sizeof(struct ccu_xfer_result);
+	}
+
+	return -EAGAIN;
+}
+
+/*
+ * ccu_poll - This function is used to give notification to the client
+ *            once transfer is completed.
+ * @filp: file pointer of device.
+ * @wait - Poll table, We can add one or more wait queues into this table.
+ *
+ * Return mask in case of success or 0 in failure.
+ */
+static __poll_t ccu_poll(struct file *filp, poll_table *wait)
+{
+	struct ccu_qup *ccu;
+
+	if (!filp || !filp->private_data)
+		return -EINVAL;
+
+	ccu = filp->private_data;
+	poll_wait(filp, &ccu->readq, wait);
+	if (atomic_read(&ccu->rx_avail)) {
+		dev_dbg(ccu->dev, "Response avilable to read\n");
+		return (POLLIN | POLLRDNORM);
+	}
+
+	return 0;
+}
+
+/*
+ * ccu_update_xfer_result - This callback function will invoke by i2c driver
+ *                          once xfer is completed.
+ * @result: pointer for struct ccu_xfer_result.
+ *
+ * Return none.
+ */
+void ccu_update_xfer_result(struct ccu_xfer_result *result)
+{
+	struct ccu_qup *ccu = ccud;
+	struct ccu_response_data *entry;
+
+	mutex_lock(&ccu->lock);
+	entry = kmalloc(sizeof(*entry), GFP_KERNEL);
+	if (!entry) {
+		dev_err(ccu->dev, "Memory allocation failed for entry\n");
+		mutex_unlock(&ccu->lock);
+		return;
+	}
+
+	entry->res_buf = kmalloc(sizeof(*result), GFP_KERNEL);
+	if (!entry->res_buf) {
+		dev_err(ccu->dev, "Memory allocation failed for buf\n");
+		kfree(entry);
+		mutex_unlock(&ccu->lock);
+		return;
+	}
+
+	memcpy(entry->res_buf, result, sizeof(*result));
+	INIT_LIST_HEAD(&entry->list);
+	/* Add received completion notification result in the list */
+	list_add(&ccu->data.list, &ccu->response_list);
+	atomic_inc(&ccu->rx_avail);
+	mutex_unlock(&ccu->lock);
+	wake_up_interruptible(&ccu->readq);
+}
+
 static int ccu_release(struct inode *inode, struct file *filp)
 {
 	struct ccu_qup *ccu;
@@ -44,6 +162,7 @@ static int ccu_release(struct inode *inode, struct file *filp)
 		return -EINVAL;
 
 	ccu = filp->private_data;
+	geni_i2c_register_callback(ccu_update_xfer_result);
 
 	return 0;
 }
@@ -691,6 +810,9 @@ static int ccu_qup_probe(struct platform_device *pdev)
 
 	spin_lock_init(&ccu->lock0);
 	spin_lock_init(&ccu->lock1);
+	init_waitqueue_head(&ccu->readq);
+	INIT_LIST_HEAD(&ccu->response_list);
+	mutex_init(&ccu->lock);
 	ccu->gpi_wq = create_singlethread_workqueue("CCU_WQ");
 	INIT_DELAYED_WORK(&ccu->get_child_dev_node, get_set_child_dev_node);
 
@@ -714,6 +836,7 @@ static int ccu_qup_probe(struct platform_device *pdev)
 		return ret;
 	}
 
+	ccud = ccu;
 	/* Queue the delayed work to be executed after 5 seconds */
 	queue_delayed_work(ccu->gpi_wq, &ccu->get_child_dev_node,
 			   msecs_to_jiffies(5000));
diff --git a/include/linux/soc/qcom/qcom_ccu_qup.h b/include/linux/soc/qcom/qcom_ccu_qup.h
index b61e78efbe03..168880d57e29 100644
--- a/include/linux/soc/qcom/qcom_ccu_qup.h
+++ b/include/linux/soc/qcom/qcom_ccu_qup.h
@@ -105,8 +105,12 @@ struct ccu_xfer_result {
  * @cfg: points to the struct ccu_qup_config.
  * @gpi_wq: workqueue to get child dev node
  * @get_child_dev_node: work function to get child dev node.
- * @lock0: spinlock to protect interrupr handler for CCU QUP0
- * @Lock1: spinlock to protect interrupr handler for CCU QUP1
+ * @lock0: spinlock to protect interrupr handler for CCU QUP0.
+ * @Lock1: spinlock to protect interrupr handler for CCU QUP1.
+ * @readq: queue of file desc waiting for read notification.
+ * @rx_avail: represent the rx data available for read.
+ * @response_list: list which represent the response.
+ * @data: pointer to the struct ccu_response_data.
  */
 struct ccu_qup {
 	struct device *dev;
@@ -132,6 +136,10 @@ struct ccu_qup {
 	struct delayed_work get_child_dev_node;
 	spinlock_t lock0;
 	spinlock_t lock1;
+	wait_queue_head_t readq;
+	atomic_t rx_avail;
+	struct list_head response_list;
+	struct ccu_response_data data;
 	u32 ccu0_irq_stat;
 	u32 ccu1_irq_stat;
 };
@@ -143,6 +151,11 @@ static int ccu_open(struct inode *inode, struct file *filp);
 static ssize_t ccu_config_write(struct file *filp, const char __user *buf,
 				size_t len, loff_t *f_pos);
 
+static ssize_t ccu_response_read(struct file *filp, char __user *buf,
+				 size_t count, loff_t *f_pos);
+
+static __poll_t ccu_poll(struct file *filp, poll_table *wait);
+
 static int ccu_release(struct inode *inode, struct file *filp);
 
 int ccu_sync_id_to_sync_ch_mapping(struct ccu_qup *ccu, struct sync_info sync[]);
-- 
2.34.1

